### PR TITLE
refactor: remove WsHttpServerKind enum and simplify server launch

### PR DESCRIPTION
**Pull request by smatthewenglish**
This PR simplifies the server launch process and removes the `WsHttpServerKind` enum. 

First pass at addressing #7491

### CODE REVIEW

File: crates/rpc/rpc-builder/src/lib.rs [hunk 1]
```diff
@@ -1634,139 +1634,48 @@ impl RpcServerConfig {
...
-                self.jwt_secret.clone(),
-                ServerKind::WS(ws_socket_addr),
-                modules.ws.as_ref().map(RpcRequestMetrics::ws).unwrap_or_default(),
-            )
-            .await?;
-            ws_local_addr = Some(addr);
-            ws_server = Some(server);
+    /// Sets up and starts a server with optional CORS, JWT authentication, and metrics at `addr`.
+    async fn launch_server(
+        addr: SocketAddr,
```
Comment (mattsse) [inline crates/rpc/rpc-builder/src/lib.rs:1639]: this needs to take `self` and use all the configured fields

and then we also need the previous logic that checks

> If both are configured on the same port, we combine them into one server.

optional layer can be configured via an optional layer

https://docs.rs/tower/latest/tower/builder/struct.ServiceBuilder.html#method.option_layer

the motivation behind this pr is to compress the previous two steps (with the intermediary enum) into one function

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/rpc/rpc-builder/src/lib.rs [hunk 2]
```diff
@@ -1961,6 +2014,15 @@ struct WsHttpServer {
     jwt_secret: Option<JwtSecret>,
 }
 
+// Define the type alias with detailed type complexity
+type WsHttpServerKind = Server<
```
Comment (smatthewenglish) [inline crates/rpc/rpc-builder/src/lib.rs:2018]: `clippy` forced me to put this type in, 'cause it said creating `WsHttpServers`s with the entire server structure was illegal

File: crates/rpc/rpc-builder/src/lib.rs [hunk 3]
```diff
@@ -1710,39 +1728,74 @@ impl RpcServerConfig {
...
+            if let Some(cors) =
+                self.http_cors_domains.take().as_deref().map(cors::create_cors_layer)
+            {
+                let cors = cors.map_err(|err| RpcError::Custom(err.to_string()))?;
+                cl = Some(cors);
+            }
+            let mut jl: Option<AuthLayer<JwtAuthValidator>> = None;
+            if let Some(secret) = self.jwt_secret.clone() {
+                jl = Some(AuthLayer::new(JwtAuthValidator::new(secret)));
+            }
```
Comment (mattsse) [inline crates/rpc/rpc-builder/src/lib.rs:1774]: this can be replaced with a `.map` on jwt_secret

File: crates/rpc/rpc-builder/src/lib.rs (hunk 2 above)
Comment (mattsse) [reply crates/rpc/rpc-builder/src/lib.rs:2018]: we want to get rid of these types entirely

**Review by mattsse — CHANGES_REQUESTED** [review]
this is great

we still need to flatten the entire launch process so we don't have any intermediary types that require large type aliases
but we can also do this separately because this will require a few more changes

so as far as this pr goes I only have some style nits, I'm fine with the type alias in this pr

File: crates/rpc/rpc-builder/src/lib.rs [hunk 4]
```diff
@@ -1681,25 +1681,43 @@ impl RpcServerConfig {
...
+            let mut jl: Option<AuthLayer<JwtAuthValidator>> = None;
+            let mut cl: Option<CorsLayer> = None;
+
+            if let Some(cors) = cors.as_deref().map(cors::create_cors_layer) {
+                let cors = cors.map_err(|err| RpcError::Custom(err.to_string()))?;
+                cl = Some(cors);
+            }
+            if let Some(secret) = jwt_secret {
+                jl = Some(AuthLayer::new(JwtAuthValidator::new(secret)));
+            }
```
Comment (mattsse) [inline crates/rpc/rpc-builder/src/lib.rs:1694]: try using `.map` here or `.and_then` so tmp vars are not needed

File: crates/rpc/rpc-builder/src/lib.rs [hunk 5]
```diff
@@ -1710,39 +1724,75 @@ impl RpcServerConfig {
...
+                .http_cors_domains
+                .take()
+                .as_deref()
+                .map(cors::create_cors_layer)
+                .and_then(|res| res.ok())
+                .or_else(|| None);
+            let jwt_auth: Option<AuthLayer<JwtAuthValidator>> = self
+                .jwt_secret
+                .as_ref()
+                .map(|secret| AuthLayer::new(JwtAuthValidator::new(secret.clone())));
```
Comment (onbjerg) [inline crates/rpc/rpc-builder/src/lib.rs:1769]: i'm seeing this cors/jwt logic almost 1:1 duplicated 3 times here, let's try to unify?

**Review by mattsse — CHANGES_REQUESTED** [review]
this is great,
we can extract the duplicated layer creation setup for auth and cors to functions so we can reuse them

**Comment by smatthewenglish** [comment]
ty for both for the feedback! I refactored the redundant code into a single function called `create_middleware`. if that's not what you had in mind i'm happy to make any changes

**Review by onbjerg — COMMENTED** [review]
just one q regarding a type and then i think this is good to go, thanks for this 😄

**Review by mattsse — APPROVED** [review]
tysm for this

I split the layer functions into standalone functions so that we don't need the alias

File: crates/rpc/rpc-builder/src/lib.rs [hunk 6]
```diff
@@ -1645,17 +1638,14 @@ impl RpcServerConfig {
...
-        &self,
-        cors: Option<String>,
-        jwt_secret: Option<JwtSecret>,
-    ) -> MiddlewareServiceBuilder {
-        let cors_layer: Option<CorsLayer> =
-            cors.as_deref().map(cors::create_cors_layer).and_then(|res| res.ok()).or_else(|| None);
-        let jwt_auth: Option<AuthLayer<JwtAuthValidator>> =
-            jwt_secret.map(|secret| AuthLayer::new(JwtAuthValidator::new(secret)));
-        tower::ServiceBuilder::new().option_layer(cors_layer).option_layer(jwt_auth)
+    /// Creates the [CorsLayer] if any
```
Comment (3x-fire-emoji) [inline crates/rpc/rpc-builder/src/lib.rs:1641]: nice! this is awesome. i also didn't like the `MiddlewareServiceBuilder` type. next time i'll know i can avoid triggering clippy by creating smaller more atomic functions

**Pull request merged**